### ISSUE TITLE
Deprecate `btowc`, `wctob`, `mbtowc`, and `wctomb`

### ISSUE THREAD

**Issue by tahonermann** [sg16-unicode/sg16#39]
`btowc` and `wctob` are unable to provide conversions for code points that require multiple code units in variable length encodings.  `mbtowc` and `wctomb` are able to handle code points that require multiple code units, but depend on global state and are therefore not thread safe.  `mbrtowc` and `wcrtomb` have superseded both.

**Comment by cubbimew** [comment]
Note WG14 has a proposal to make mbtowc and wcromb thread-safe: [n2281](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2281.htm).
[Minutes](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2307.pdf) say "Agree with the goal in principle" and "The paper needs more work"

**Comment by tahonermann** [comment]
@cubbimew Thank you for that link!

**Comment by cubbimew** [comment]
If btowc/wctob can be considered for deprecation, what about implicit conversions between `char` and `wchar_t`? Those are even more broken (they are basically btowc/wctob minus error-checking) and more sneaky (e.g. they let you construct a garbage `wstring` from a `string` iterator pair)

**Comment by tahonermann** [comment]
I agree those conversions are more broken.  I don't know how realistic it would be to make changes for `wchar_t` (and `char16_t` and `char32_t`), but it would be nice it we could.  I've been leaning more towards leaving code unit/code point types as is and introducing new "character" types that are not integral types and do not participate in the traditional integral promotions/conversions.
